Update Chapel package and add variants

arezaii opened the pull request:
This PR updates the Chapel package definition to bring the supported versions up to date and add variants for most configuration options. 

It also adds post-install steps for various levels of testing that the installation went as expected.

arezaii committed:
add variants and build options to chapel package

arezaii committed:
add cray detection taken from upcxx

arezaii committed:
refactor elements of chapel package.py

arezaii committed:
fix formatting errors revealed by CI

arezaii committed:
reformatting to pass spack style

arezaii committed:
additional variants based on CHPL_ env vars

arezaii committed:
add post install checks

mppf commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,385 @@
+
+    variant(
+        "llvm",
+        default="unset",
```
Doesn't a default of `CHPL_LLVM=unset` mean that it won't build unless you provide one? Would the default being `spack` would be more reasonable here than making it required?

arezaii replied:
I have been wrestling with reasonable defaults for some of these. The pain of specifying spack by default is that it will likely have to build from source. It might end up being the best default we can offer though.

mppf commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,385 @@
+        "gmp",
+        description="Build with gmp support",
+        default="unset",
+        values=("unset", "system", "none", "bundled"),
```
Would using a Spack GMP be a reasonable default for this one?

arezaii replied:
I think spack would be happiest if we just used their packages over our bundled software in most cases. It looks like I can add a spack option here.

arezaii replied:
My initial attempts to add spack GMP is giving me problems, not finding `gmp.h` unless I add it to `CPATH`, then complaining about not finding `libgmp` even though I have added its `lib/` to `LD_LIBRARY_PATH` and confirmed `libgmp.so` and `.a` exist there. I'm still working on it, just noting it here along with another thing - the GMP spack package is "currently looking for a maintainer" and is back at 6.2.1 (though I don't necessarily think that's the problem or an incompatibility, just a note)

mppf replied:
In that case I would think that leaving it unset is reasonable for now. IIRC, that will use a system GMP if we find one that is suitable, and if not, use the bundled GMP.

mppf commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,385 @@
+    )
+
+    variant(
+        "aux_filesys",
```
Pretty sure this is deprecated and doesn't need to be in the Spack package here

arezaii replied:
Oh, I didn't see that it was marked deprecated, our environment setup docs might need an update in that case.

mppf commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,385 @@
+
+    variant(
+        "locale_model",
+        values=("flat", "numa"),
```
NUMA locale model is deprecated but the GPU locale model is of current interest

arezaii replied:
Oops, I'll remove numa. I plan to work on GPU support in the coming sprint

arezaii committed:
update variant names, make llvm and gmp default to spack, remove numa locale

bonachea commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,398 @@
+    def setup_chpl_comm(self, env, spec):
+        env.set("CHPL_COMM", spec.variants["comm"].value)
+        if spec.variants["substrate"].value != "none":
+            env.set("CHPL_COMM_SUBSTRATE", spec.variants["substrate"].value)
```
`"substrate"` above should be `"comm_substrate"`

arezaii committed:
update substrate variant name in env logic

arezaii committed:
start of adding gpu enabled builds with cuda

arezaii committed:
fixup style

arezaii committed:
experiment to add amd gpu support

arezaii committed:
fix requires condition

arezaii committed:
add launcher variant and package_module testing

arezaii committed:
add optional tests for some package modules

arezaii committed:
update comments, minor formatting

arezaii committed:
enable make check as variant with +check

arezaii committed:
sort variants, values by name

arezaii committed:
additional variants, move comments, ignore windows

arezaii commented:
Hi spack team, I am looking for a reviewer to help shepherd these changes into our package. 
I am tagging some members I see on recent PRs here just to get some visibility and maybe more guidance- @haampie @alalazo - thank you!

arezaii committed:
fix bad punctuation chars

arezaii committed:
replace custom gpu variants with spack build systems for cuda, rocm

arezaii committed:
try to respect spack compiler arg

arezaii committed:
fix spack style errors

arezaii committed:
update gpu variants, specifically for amd

arezaii committed:
update style, rebase on develop

arezaii committed:
updates to prepare for Chapel 2.0 release

arezaii committed:
adjust version numbers based on feedback

arezaii committed:
attempt to fixup testing

arezaii committed:
remove some troublesome testing

arezaii committed:
fix missing python range

arezaii committed:
fix spack style errors

arezaii commented:
`spack test run chapel` is giving me trouble when using `+cuda` and the `CudaPackage` build system class. It seems like `nvcc` isn't being made available during the test ran with `spack test run chapel`. If I manually load `chapel+cuda` or just `cuda` before the test, it works, but just running the test fails. Using `which` to find `nvcc` during the test indicates `nvcc` isn't available.

arezaii committed:
add comment about why make check fails at build time

arezaii committed:
add ugni comm option, map spack target to chpl_cpu

arezaii committed:
update llvm versions

arezaii committed:
remove older versions, chpldoc conflicts

arezaii committed:
add hpe-cray-ex detection, target cpu options, remove target and host compiler variants

arezaii committed:
add patch for cray prgenv to avoid spack wrapper

arezaii committed:
add dep on binutils+headers when using bundled llvm

bonachea committed:
Enforce that comm=gasnet requires comm_substrate= set to the appropriate network

bonachea committed:
Don't actually need to explicitly set CHPL_COMM_SUBSTRATE

This is handled by first line in `setup_env_vars()`

bonachea commented:
@spackbot fix style

bonachea committed:
Fix style

bonachea committed:
Set CHPL_TARGET_{CC,CXX} to avoid the Spack wrappers

These compilers need to be invokable post-install during `chpl` execution, for at least some configurations.

No need to set CHPL_HOST_{CC,CXX}, as these *should* never be invoked post-install.

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,773 @@
+        default="spack",
+        description="LLVM backend type. Use value 'spack' to have spack "
+        "handle the LLVM package",
+        values=("bundled", "none", "spack", "system"),
```
Generally, because Spack supports externals (use of system packages), IMO it would be good to remove the "system" option (that is covered by "spack"+external).

arezaii replied:
so the recommendation is that users would need to have spack detect their system LLVM and provide it as an external package? I think that is fine, this is really a speed/convenience feature to avoid having to compile LLVM for 30+ minutes when the bundled or spack variants are selected on a new system (without an LLVM spackage installed already)

bonachea replied:
> so the recommendation is that users would need to have spack detect their system LLVM and provide it as an external package? 

@arezaii : the `llvm=spack` option already serves two use cases:
1. An appropriate LLVM is available on the system and was detected as an external (e.g. using `spack external find llvm`)
2. LLVM is not built yet, but we want Spack to build it. 
    * This option is notably better than a Chapel-bundled build because it allows one to quickly install multiple `chapel` variants all backed by the same `llvm` build.

So I agree with @scheibelp, we should remove the superfluous `llvm=system` option.

I suspect the same logic implies we should remove `gmp=system` in favor of `gmp=spack`.

We should similarly also consider replacing all of `host_jemalloc=system` and `libfabric=system` and `unwind=system` with `=spack` variants. The Chapel CMake will still treat these "system" installs, but they are all supported by Spack packages that can be either built from source or registered as externals.

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,773 @@
+
+    for opt, dep in package_module_dict.items():
+        depends_on(dep, when="package_modules={0}".format(opt), type=("run", "build", "link"))
+        depends_on(dep, when="package_modules=all", type=("run", "build", "link"))
```
We generally want to avoid redundant variant values. While I understand the convenience of being able to specify `=all`, it is preferable to have individual variants for each component that would be enabled.

arezaii replied:
ok, I worried about that. So IIUC in this case I should refactor each value (except `all`) in `package_modules` into a boolean variant of its own.

scheibelp replied:
Yes

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,773 @@
+    depends_on("gmp", when="gmp=spack", type=("build", "link", "run"))
+
+    depends_on("python@3.7:")
+    depends_on("cmake@3.16:")
```
(minor) I don't see other mentions of `cmake`, do later Chapel versions use a CMake-based build?

arezaii replied:
cmake is used to build the chapel compiler and LLVM when the bundled option is selected

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,773 @@
+        if self.compiler_map.get(self.spec.compiler.name) is None:
+            raise InstallError(
+                "Chapel did not recognize the {0} compiler".format(self.spec.compiler.name)
+            )
```
I'd recommend encoding this as

```
requires("%compiler1", "%compiler2", policy="one_of")
```

which would enforce this at concretization time vs. install time.

bonachea committed:
Enable Spack to provide GASNet

This enables a Spack user to override the GASNet version to be used.

For example, this command pulls in the latest release of GASNet-EX,
which is newer than the version embedded in the 2.0.0 release:

```
spack install chapel@2.0.0 comm=gasnet comm_substrate=ofi gasnet=spack ^gasnet@master
```

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,784 @@
+            env.prepend_path("LD_LIBRARY_PATH", self.spec["curl"].prefix.lib)
+            env.prepend_path("PKG_CONFIG_PATH", self.spec["curl"].prefix.lib.pkgconfig)
+
+        if self.spec.variants["cuda"].value:
```
(minor) the idiomatic way to do this is `if self.spec.satisfies("+cuda")`

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,784 @@
+            env.prepend_path("CPATH", self.spec["libzmq"].prefix.include)
+            env.prepend_path("LD_LIBRARY_PATH", self.spec["libzmq"].prefix.lib)
+            env.prepend_path("PKG_CONFIG_PATH", self.spec["libzmq"].prefix.lib.pkgconfig)
+            env.prepend_path("PKG_CONFIG_PATH", self.spec["libsodium"].prefix.lib.pkgconfig)
```
We should be adding these directories automatically (to `PKG_CONFIG_PATH`): did you get an error without this?

arezaii replied:
all of these changes (i.e. those touching env pathing vars) were made after running into problems when the additions were not present. After some discussion with @bonachea last night, we believe there's a challenge to make the chapel language and compiler work as a spack package because the dependencies that spack supplies are relevant to not only chapel, but to the applications that chapel builds. Currently, we do not encode `RPATH`s for all the dependencies in the executable and so when these compiled binaries try to run, they lack the necessary linking information.

bonachea replied:
To restate @arezaii 's point more directly: 
The chapel package we are building is itself a compiler, and long after `spack install` creates the `chpl` binary, end users "use" Chapel by invoking the `chpl` compiler to compile and link their applications written in the Chapel language. 

The execution of the `chpl` compiler requires calls to the system C compiler and linker, which cannot automatically benefit from the spack compiler wrappers or other spack magic because these invocations happen long after `spack install` is complete. Also, we believe the `chpl` compiler does not (yet) correctly pass RPATHs to the application link step corresponding to libraries built using Spack; the `chpl` application link step instead mistakenly assumes that any such libraries (detected as pre-existing "system" libraries during Chapel configure) are part of the default system linker search path. We hope to eventually improve logic in the Chapel compiler to resolve that latter misbehavior, but that's not a trivial change and won't help the current releases of Chapel.


The insertion of all these `LD_LIBRARY_PATH` settings in `setup_run_environment()` primarily ensures that end users can `spack load chapel` (or load a Spack-generated module) to get the necessary library paths to run their Chapel-built applications. We recognize this is a non-ideal user experience, as ideally end users of a Chapel install should never need to know or care that Spack was used by their sysadmin to install Chapel. However it provides a working solution today for current Chapel releases until the Chapel compiler can be improved to integrate better with Spack-provided dependencies. 

I'm less confident that the `CPATH` and `PKG_CONFIG_PATH` are strictly necessary in `setup_run_environment()`, but they also don't seem harmful.

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,784 @@
+            env.prepend_path("CPATH", self.spec["libyaml"].prefix.include)
+
+        if "zmq" in self.get_package_modules:
+            env.prepend_path("CPATH", self.spec["libzmq"].prefix.include)
```
(note: problems have occurred when combining CPATH updates w/ `pkgconfig`: did you find you have to set both?)

bonachea replied:
We are now slightly more cautious in prepending things to `CPATH`, in particular to ensure we never prepend `/usr/include` as a result of a distro-provided external.

scheibelp replied:
FWIW there is a utility function available for filtering these: `spack.util.environment.filter_system_paths`

bonachea replied:
I just changed the logic in question to use `spack.util.environment.is_system_path` as you suggest.

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,784 @@
+
+    conflicts("platform=windows")  # Support for windows is through WSL only
+
+    conflicts("rocm", when="cuda", msg="Chapel must be built with either CUDA or ROCm, not both")
```
This has to be 

```
conflicts("+rocm", when="+cuda")
```

(i.e., tests for boolean variants have to use `+` or `~`)

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,784 @@
+            env.unset(var)
+
+    def configure(self, spec, prefix):
+        configure("--prefix={0}".format(prefix))
```
Note: this is the default implementation of `configure()` provided by inheriting `AutotoolsPackage`

arezaii replied:
Since we added other things here, I think we need to preserve this to avoid overriding the base without executing `configure`.

bonachea replied:
In 2f9a35c36a I went ahead and moved `setup_gasnet()` to `@run_before("configure")`, allowing this override to be deleted.

scheibelp commented on var/spack/repos/builtin/packages/chapel/fix_spack_cc_wrapper_in_cray_prgenv.patch:
```diff
@@ -0,0 +1,82 @@
++        real_cc = os.path.join(os.environ["CRAYPE_DIR"], "bin", "cc")
+         # Get compilation arguments
+-        opts = run_command(['cc', '--cray-print-opts=cflags'])
++        opts = run_command([real_cc, '--cray-print-opts=cflags'])
```
I'm not sure to what extent this is used in the build, but FWIW the compiler wrappers in Spack are responsible for inserting RPATHs, so bypassing them can lead to run-time errors (without LD_LIBRARY_PATH manipulations, a few of which I see in this PR).

bonachea replied:
@scheibelp I don't really understand this feedback. 

Your comment is on a patch that fixes a defect in the Chapel's underlying configure logic, which currently makes the questionable assumption that the compiler used on a Cray EX is always the Cray PrgEnv wrapper. This is almost always the case for manual builds of Chapel, but Spack's compiler handling breaks this assumption.

This patch has (almost) nothing to do with RPATH or LD_LIBRARY_PATH: it's about avoiding a hard error in the Chapel build when `$CC` is actually a Spack compiler wrapper that rejects the special options provided by the Cray PrgEnv wrapper.

scheibelp replied:
It looked like this was replacing a call like `cc ...` with `/abspath/to/cc`, which I thought would bypass Spack's compiler wrapper. I didn't know what context this runs in (e.g. during the build, or after), but this could in particular be an issue if run during the build.

I noticed other parts of the PR setting LD_LIBRARY_PATH: that should not be required if the Spack compiler wrappers are used because they insert RPATHs. This was the only part of the PR that I saw directly interfering with compiler paths, but on closer examination it only seems to be changing that path for this single call, and not for other calls (e.g. when running `cc` to build objects).

>  it's about avoiding a hard error in the Chapel build when $CC is actually a Spack compiler wrapper that rejects the special options provided by the Cray PrgEnv wrapper.

based on this it sounds like I was mistaken and generally the build is using the Spack compiler wrappers (which is good, although that rekindles my confusion over spots that set `LD_LIBRARY_PATH`). The Spack compiler wrapper *should* be willing to pass arbitrary options through to the underlying compiler, so

* this seems like a bug in Spack's compiler wrapper
* I'm curious what error you got when running this without the patch

bonachea replied:
> This was the only part of the PR that I saw directly interfering with compiler paths, ...
> generally the build is using the Spack compiler wrappers (which is good, although that rekindles my confusion over spots that set LD_LIBRARY_PATH).

This is not entirely true. A recent commit (which might post-date your reviewed version) explicitly bypasses the Spack compiler wrapper for the target compiler, because it needs to be usable in some cases when the Chapel compiler is invoked by the end-user after Spack installation. 

However I suspect this patch is still necessary to ensure the Host compiler (which remains the Spack wrapper) is handled correctly by Chapel configure.

> > it's about avoiding a hard error in the Chapel build when $CC is actually a Spack compiler wrapper that rejects the special options provided by the Cray PrgEnv wrapper.
> 
> based on this it sounds like I was mistaken and generally the build is using the Spack compiler wrappers (which is good, although that rekindles my confusion over spots that set `LD_LIBRARY_PATH`). The Spack compiler wrapper _should_ be willing to pass arbitrary options through to the underlying compiler, so
> 
> * this seems like a bug in Spack's compiler wrapper
> * I'm curious what error you got when running this without the patch

It's not that Spack's compiler wrapper is failing to pass the option, it's that Spack's compiler wrapper happily passes it to the underlying compiler (e.g. raw gcc) which is not Cray's compiler wrapper that unmodified Chapel configure expects on a Cray system. This leads to a failure inside the Chapel build that looks like:
```
gcc: error: unrecognized command-line option '--cray-print-opts=cflags'
```

IOW if there's a bug it's that Chapel assumes the C compiler on a Cray is always the Cray PrgEnv wrapper, and Spack automatically "unwraps" the Cray PrgEnv wrapper and targets the underlying native compiler directly (but I'm pretty sure that's by design).

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,784 @@
+        "to the given architecture when using --specialize (and --fast).",
+        default="unset",
+        multi=False,
+    )
```
I don't see this variant being used, and it seems like most possible values would be covered by `spec.target`.

arezaii replied:
this is another value that is just used inside the chapel build scripts, but it is important to specify when installing because of the heterogenous nature of some machines

bonachea committed:
Add version 2.0.1

bonachea committed:
Fix to ensure we never add /usr to CPATH

Turns out that adding "/usr" to `$CPATH` (e.g. due to a spack external
dependency) breaks the LLVM build in Horrible Ways with `llvm=bundled`, eg:

```
/usr/include/c++/12/cstdlib:75:15: fatal error: stdlib.h: No such file or directory
   75 | #include_next <stdlib.h>
      |               ^~~~~~~~~~
compilation terminated.
```
Make sure we never do that.

bonachea committed:
fix_spack_cc_wrapper_in_cray_prgenv.patch remains necessary

Testing on Perlmutter with `llvm-bundled` shows this remains necessary
through at least current chapel:main at 57c5b877a2

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,811 @@
+        default="unset",
+        values=platform_opts,
+        multi=False,
+    )
```
I only see logic that checks if this is "unset", is there a reason to have other values?

arezaii replied:
the user has an option to choose a `host_platform` and we'll use it if they do, but if they don't we also want to detect that to allow for the chapel build scripts to do their auto-detection based on the unset variable.

scheibelp replied:
> the user has an option to choose a host_platform and we'll use it if they do

(minor) I might be confused, because I don't see any logic here that uses any value other than `unset`, which is to say the following variant definition would be equivalent:

```
    variant(
        "host_platform",
        description="Host platform",
        default="unset",
        values=("set", "unset"),  # no need to define a number of host platforms here
        multi=False,
    )
```

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,811 @@
+    #     values=compilers,
+    #     description="Compiler suite for building the Chapel compiler on CHPL_HOST_PLATFORM",
+    #     default="unset",
+    # )
```
Generally speaking, it is also recommended to create definitions for any compilers that are missing, as trying to use a compiler outside of what concretization assigned can lead to problems (e.g. Spack's compiler wrappers would point to the compiler it chose, so you could end up with some objects built with one compiler and some built by another).

Some of the commented sections would make sense to keep, since I think they lay the foundations for future work; in this case IMO it would make sense to remove this bit (and the associated `compilers` set, which appears to only be defined for this).

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,811 @@
+    variant(
+        "host_jemalloc",
+        values=("bundled", "none", "system", "unset"),
+        default="unset",
```
(question) what's the difference between `"none"` and `"unset"`?

arezaii replied:
the reason for most of these including 'unset' is because the chapel build system includes a lot of python scripts to detect things in the environment if they weren't explicitly set by the user. In this case, there's some logic in those scripts that treats an unset `CHPL_HOST_JEMALLOC` variable differently than 'none', where the former indicates default to system on darwin hosts, and default to bundled on linux hosts, unless the HOST_MEM was set to 'cstdlib'.
https://chapel-lang.org/docs/usingchapel/chplenv.html#chpl-host-jemalloc

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,811 @@
+        # and just rely on minimal hello world and version number tests.
+        # We can't just run make check here because chpl isn't yet installed
+        # and we need it to be in the PATH for make check to work
+        pass
```
(minor) It might be worth reorganizing/shortening:

* By default all `AutotoolsPackage`'s run `make check` and `make test` as part of the `check` phase
* You are overriding it to skip those because
  * `make check` doesn't work at build time
  * `make test` is expensive

scheibelp reviewed (changes requested):
Notes:

* Comments prefixed with (minor) can be ignored depending on preference (they are suggestions)
* Main requests are:
  * Refactor of `package_modules` variant
  * Concerns about management of CPATH, LD_LIBRARY_PATH, and PKG_CONFIG_PATH (mainly that Spack should be handling these already for you except in some cases when compiler wrappers are bypassed)

bonachea committed:
Update is_CrayEX predicate

See #44618

bonachea committed:
update maintainers

arezaii committed:
remove executables to avoid presenting installs to spack external

arezaii committed:
use spec.satisfies and '+<dep>' for boolean evals, per reviewer feedback

arezaii committed:
fixup comments describing why we skip self check step

arezaii committed:
remove step in configure as it was redundant to the default

bonachea committed:
Add `hwloc=spack`

This utilizes a currently undocumented `CHPL_HWLOC=system` setting,
which nevertheless appears to work.

bonachea committed:
fix_spack_cc_wrapper_in_cray_prgenv.patch: Revert accidentally committed patch hunks

bonachea committed:
fixup: Add `hwloc=spack`

Add missing depends_on

bonachea committed:
Refine CUDA dependencies for llvm=spack

bonachea committed:
fix style

arezaii committed:
update LLVM variant to remove system, update description

arezaii committed:
Revert "remove step in configure as it was redundant to the default"

This reverts commit 5a982e99afc3706b4d504c91ea38fd9f97b746be.

bonachea committed:
hwloc=spack variant is only functional/supported for @2.1:

bonachea committed:
Express setup_gasnet() in a more idiomatic way

and remove the now-superfluous configure() override

bonachea commented:
@spackbot fix style

arezaii committed:
remove compilers list and commented variants

arezaii committed:
refactor package_modules into variants, default true for all

arezaii committed:
fixup bad variable 'name' in package

bonachea committed:
Chapel 2.1+ requires GASNet-EX 2024.5+ for Chapel PR #25140

arezaii committed:
add 'requires' entry for compiler names

arezaii committed:
change 'system' deps to 'spack' deps, add comments, remove duplicated env setting

bonachea committed:
Add Chapel developer mode

Useful when reporting build errors to Chapel maintainers

bonachea committed:
Check spack.util.environment.is_system_path when manipulating CPATH

arezaii committed:
move more deps from system to spack

arezaii committed:
widen available LLVM version compatibility to match chapel releases

arezaii committed:
add depends_on for jemalloc when spack dep selected

arezaii committed:
add host_jemalloc conflicts to preempt failure in chpl_jemalloc.py

bonachea committed:
Ensure that gmp and hwloc are visible in PKG_CONFIG_PATH for test environment

By default Spack only seems to add these to PKG_CONFIG_PATH for the build-env,
despite these being clearly marked as "test" dependencies.
Also need the "libpciaccess" dependency or pkg-config fails to recognize "hwloc"

bonachea committed:
Add support for comm=gasnet comm_substrate=smp

bonachea committed:
Force TARGET_{CC,CXX} to LLVM when using GPU support

This is a (poorly documented) requirement enforced by the Chapel makefiles

arezaii committed:
add paths for dependency packages to system variables at load time

arezaii committed:
remove unrecognized compiler name definitions

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,813 @@
+        "unwind",
+        description="Build with unwind library for stack tracing",
+        default="none",
+        values=("bundled", "none", "spack"),
```
(minor) I don't see the logic handling a distinction between `none` and `bundled` for this particular variant.

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,813 @@
+        default="unset",
+        values=platform_opts,
+        multi=False,
+    )
```
(minor) I don't see this variant being used anywhere like (unlike `host_platform`, for example), likewise for `target_cpu`/`target_arch`/`timers`/`launcher`.

In https://github.com/spack/spack/pull/42197/files#r1631482021 you mention:

> this is another value that is just used inside the chapel build scripts, but it is important to specify when installing because of the heterogenous nature of some machines

what I mean is that while you unset the `"CHPL_TARGET_ARCH"` env variable, I do not see anywhere in this package that reads the value of this variant, so it cannot have an impact on the build.

arezaii replied:
I think what I didn't explain very well was that all of the keys are defined in `chpl_env_vars` and the processing around line 561 evaluates each variant and sets the value in the environment if the variant was given anything other than 'unset' so that Chapel's build scripts can detect them and that will affect the build.

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,813 @@
+        "libfabric",
+        default="unset",
+        description="When building with ofi support, specify libfabric option",
+        values=("bundled", "spack", "unset"),
```
(minor) Similar theme to other comments in this review, there is no distinct handling of `bundled` and `unset`, so the install behavior would be the same for each of these.

Analogous comment for `none`/`bundled` for `hwloc`; likewise `gmp`.

scheibelp commented on var/spack/repos/builtin/packages/chapel/package.py:
```diff
@@ -3,24 +3,813 @@
+        values=("array_on_device", "unified_memory"),
+        default="array_on_device",
+        multi=False,
+    )
```
(minor) similar theme to other comments: while I see you unsetting the `CHPL_GPU_MEM_STRATEGY` env variable, nothing ever looks at this variant to make a decision, so it is not currently impacting the outcome of the install.

scheibelp reviewed (dismissed):
My remaining concerns are about unused variants (or similarly, variants with redundant values).

I'm fine with merging this although I wanted to make sure we agreed about that in case I missed something (once you respond, I'll collapse whatever comments remain and merge)

bonachea commented:
> My remaining concerns are about unused variants (or similarly, variants with redundant values).
> 

@scheibelp Thanks for the review. I think what you're missing is these lines:

```python
    def setup_env_vars(self, env):
        for v in self.spec.variants.keys():
            self.setup_if_not_unset(env, "CHPL_" + v.upper(), self.spec.variants[v].value)
```

which have the effect of automatically transliterating each variant `val=var` to `CHPL_VAL=var`.

I believe that answers all your questions about seemingly unused variants?

scheibelp commented:
> which have the effect of automatically transliterating each variant val=var to envvar CHPL_VAL=var, while converting "spack" values to "system" (and "unset" to no setting) as expected by the Chapel build infrastructure.
> I believe that answers all your questions about seemingly unused variants?

I think most of it, but:

* I'm not sure if it directly addresses https://github.com/spack/spack/pull/42197#discussion_r1643426686 (unless `none` and `bundled` are distinct values that you want to pass to CHPL_UNWIND
* Does this set `CHPL_*` environment variables which don't have any meaning for Chapel? It looks like for example that it would set an env variable like `CHPL_CHPLDOC`?
  * IMO it would make it easier to review/edit in the future if this list of auto-translated variants were listed out explicitly in `setup_env_vars`

arezaii committed:
add pathing required for zmq programs to build and run

arezaii committed:
add comment explaining seemingly unused variants

bonachea commented:
> * I'm not sure if it directly addresses [Update Chapel package and add variants #42197 (comment)](https://github.com/spack/spack/pull/42197#discussion_r1643426686) (unless `none` and `bundled` are distinct values that you want to pass to CHPL_UNWIND

@scheibelp Here is the [documentation for CHPL_UNWIND](https://chapel-lang.org/docs/usingchapel/chplenv.html#chpl-unwind). 
As you hypothesize, `=bundled`, `=system` (aka `=spack`), and `=none` all activate distinct behaviors in the Chapel build system, corresponding to roughly "build libunwind as part of Chapel build", "use the Spack-provided libunwind" and "disable support for libunwind", respectively.  Many of the other variants have similar three options.

> * Does this set `CHPL_*` environment variables which don't have any meaning for Chapel? It looks like for example that it would set an env variable like `CHPL_CHPLDOC`?

I agree this is probably the case, but should also be harmless because the Chapel build system will just ignore environment variables it doesn't recognize.

>   * IMO it would make it easier to review/edit in the future if this list of auto-translated variants were listed out explicitly in `setup_env_vars`

The current approach has the advantage of being more factored and concise, although I agree it can potentially lead to misundertandings. @arezaii just added a comment to the line-in-question to hopefully improve readability for future reviewers. Does that satisfy your concern, or do you think it's important to re-engineer this before merging this revision? 

I think it's also worth noting there's a new Chapel release pending in a few weeks, so there will definitely be a follow-up Spack PR in a few weeks to add that and possibly other enhancements. In the meantime, this PR provides a package that effectively doesn't exist in a usable form in spack:develop.

scheibelp commented:
> Does that satisfy your concern,

Yes - thanks!

arezaii commented:
> * I'm not sure if it directly addresses [Update Chapel package and add variants #42197 (comment)](https://github.com/spack/spack/pull/42197#discussion_r1643426686) (unless `none` and `bundled` are distinct values that you want to pass to CHPL_UNWIND

yes, those are distinct values that we want to pass to `CHPL_UNWIND` per https://chapel-lang.org/docs/usingchapel/chplenv.html#chpl-unwind

> Does this set CHPL_* environment variables which don't have any meaning for Chapel? It looks like for example that it would set an env variable like CHPL_CHPLDOC?

yes, although we'll ignore them, one way that can be addressed is by filtering the variants based on the `chpl_env_vars` list

> IMO it would make it easier to review/edit in the future if this list of auto-translated variants were listed out explicitly in setup_env_vars

Dan beat me to it, but I agree with his sentiment on this, including that we _can_ change it if need be.

scheibelp commented:
Does the following commit message seem reasonable?

```
* add cray detection taken from upcxx
* add CUDA/ROCm support
* add numerous pass-through options to Chapel build,
  like gpu_mem_strategy, comm_substrate, etc.; all variants are
  translated to analogous CHPL_* environment variables. As a side
  effect, this defines a number of environment variables that are
  not actually used by Chapel.
* Define LD_LIBRARY_PATH, LIBRARY_PATH, and PKG_CONFIG_PATH to
  help programs built with Chapel properly locate needed runtime
  dependencies
```

arezaii commented:
> Does the following commit message seem reasonable?
> 
> ```
> * add cray detection taken from upcxx
> * add CUDA/ROCm support
> * add numerous pass-through options to Chapel build,
>   like gpu_mem_strategy, comm_substrate, etc.; all variants are
>   translated to analogous CHPL_* environment variables. As a side
>   effect, this defines a number of environment variables that are
>   not actually used by Chapel.
> * Define LD_LIBRARY_PATH, LIBRARY_PATH, and PKG_CONFIG_PATH to
>   help programs built with Chapel properly locate needed runtime
>   dependencies
> ```

I am happy with this, thank you @scheibelp!

bradcray commented:
Hooray!  Thanks for all the hard work here, @arezaii and @bonachea, and for the advice and feedback @scheibelp (and @bonachea)!
